lib/compiler/src/beam_doc.erl

```diff
@@ -0,0 +1,285 @@
+update_meta(#docs{meta = Meta0}=State, Meta1) ->
+    State#docs{meta = maps:merge(Meta0, Meta1)}.
+
+-spec update_doc(State :: internal_docs(), Doc :: unicode:chardata() | atom()) -> internal_docs().
+update_doc(State, Doc) ->
+State#docs{doc = string:trim(Doc)}.
+
+-spec update_module(State :: internal_docs(), ModuleName :: unicode:chardata()) -> internal_docs().
+update_module(#docs{}=State, ModuleName) ->
+    State#docs{module_name = ModuleName}.
+
+-spec update_export_funs(State :: internal_docs(), proplists:proplist()) -> internal_docs().
+update_export_funs(State, ExportedFuns) ->
+    ExportedFuns1 = sets:union(State#docs.exported_functions, sets:from_list(ExportedFuns)),
+    State#docs{exported_functions = ExportedFuns1}.
+
+-spec update_export_types(State :: internal_docs(), proplists:proplist()) -> internal_docs().
+update_export_types(State, ExportedTypes) ->
+    ExportedTypes1 = sets:union(State#docs.exported_types, sets:from_list(ExportedTypes)),
+    State#docs{exported_types = ExportedTypes1}.
+
+update_export_all(State, ExportAll) ->
+    State#docs{ export_all = ExportAll }.
+
+extract_documentation([{attribute,_ANNO,compile, export_all} | T]=_AST, State) ->
+    extract_documentation(T, update_export_all(State, true));
+extract_documentation([{attribute,_ANNO,export,ExportedFuns} | T]=_AST, State) ->
+    extract_documentation(T, update_export_funs(State, ExportedFuns));
+extract_documentation([{attribute,_ANNO,export_type,ExportedTypes} | T]=_AST, State) ->
+    extract_documentation(T, update_export_types(State, ExportedTypes));
+extract_documentation([{attribute, _Anno, file, {ModuleName, _A}} | T], State) ->
+    extract_documentation(T, update_module(State, ModuleName));
+extract_documentation([{attribute, _Anno, doc, Meta0}=_AST | T], State) when is_map(Meta0) ->
+    extract_documentation(T, update_meta(State, Meta0));
+extract_documentation([{attribute, _Anno, doc, Doc}=_AST | T], State) when Doc =:= hidden orelse Doc =:= false ->
+    extract_documentation(T, State);
+extract_documentation([{attribute, _Anno, doc, Doc}=_AST | T], State) when is_list(Doc) ->
+    extract_documentation(T, update_doc(State, Doc));
+extract_documentation([AST0 | _T]=AST,
+                      #docs{meta = #{ equiv := {call,_,{atom,_,EquivF},Args}} = Meta}=State)
+    when is_tuple(AST0) andalso (tuple_size(AST0) > 2 orelse tuple_size(AST0) < 6) ->
+    Meta1 = Meta#{ equiv := {EquivF, length(Args)}},
+    extract_documentation(AST, update_meta(State, Meta1));
+extract_documentation([{function, Anno, F, A, [{clause, _, ClauseArgs, _, _}]}=_AST | T],
+                      #docs{exported_functions = ExpFuns}=State) ->
+    maybe
+        true ?= sets:is_element({F, A}, ExpFuns) orelse State#docs.export_all,
+        FunDoc = template_gen_doc({function, Anno, F, A, ClauseArgs}, State),
+        [FunDoc | extract_documentation(T, reset_state(State))]
+    else
+      _ ->
+            extract_documentation(T, reset_state(State))
+    end;
```

Comment:
When a `maybe` only uses a single `?=`, I recommend using a plain `case`:

```suggestion
    case sets:is_element({F, A}, ExpFuns) orelse State#docs.export_all of
        true ->
            FunDoc = template_gen_doc({function, Anno, F, A, ClauseArgs}, State),
            [FunDoc | extract_documentation(T, reset_state(State))]
        false ->
            extract_documentation(T, reset_state(State))
    end;
```

Reply:
Thanks for the advice. After doing the changes to a `case` clause, I see why that makes sense.
Thanks